service/src/test/kotlin/app/cash/backfila/service/BackfillRunnerTest.kt

```diff
@@ -589,6 +588,161 @@ class BackfillRunnerTest {
+            Result.success(RunBatchResponse.Builder().build()))
+
+        waitForOtherCoroutines()
+        var firstBatchCursor = getSinglePartitionCursor(runner)
+        checkNotNull(firstBatchCursor) // cursor was set when the first batch succeeded
+
+        // return a remaining batch
+        val newStart = initialRequest.batch_range.start.utf8().toInt() + 1
+        val remainingRange = initialRequest.batch_range.newBuilder()
+            .start(newStart.toString().encodeUtf8())
+            .build()
+        fakeBackfilaClientServiceClient.runBatchResponses.send(
+            Result.success(RunBatchResponse.Builder().remaining_batch_range(remainingRange).build()))
+
+        waitForOtherCoroutines()
+        assertThat(getSinglePartitionCursor(runner)).isEqualTo(firstBatchCursor) // cursor is unchanged
+        fakeBackfilaClientServiceClient.runBatchResponses.send(
+            Result.success(RunBatchResponse.Builder().build()))
+
+        waitForOtherCoroutines()
+        assertThat(getSinglePartitionCursor(runner)).isEqualTo(firstBatchCursor) // cursor is unchanged
+
+        // Only two requests are sent. The first is in response to the first batch succeeding and
+        // the second is the follow up request
+        val queuedRequest = fakeBackfilaClientServiceClient.runBatchRequests.receive()
+        assertThat(queuedRequest).isNotNull
+        val followupRequest = fakeBackfilaClientServiceClient.runBatchRequests.receive()
+        assertThat(remainingRange).isEqualTo(followupRequest.batch_range)
+
+        waitForOtherCoroutines()
+        assertThat(fakeBackfilaClientServiceClient.runBatchRequests.poll()).isNull()
+
+        // complete the two stuck requests currently being processed
+        fakeBackfilaClientServiceClient.runBatchResponses.send(
+            Result.success(RunBatchResponse.Builder().build()))
+        waitForOtherCoroutines()
+        assertThat(getSinglePartitionCursor(runner)).isEqualTo(firstBatchCursor) // cursor is still unchanged
+        fakeBackfilaClientServiceClient.runBatchResponses.send(
+            Result.success(RunBatchResponse.Builder().build()))
+        waitForOtherCoroutines()
+        val cursor = getSinglePartitionCursor(runner)
+        assertThat(cursor).isGreaterThan(firstBatchCursor) // cursor increased
+        assertThat(cursor).isEqualTo(queuedRequest.batch_range.end.utf8().toLong())
+
+        // And we can get another 3
+        assertThat(fakeBackfilaClientServiceClient.runBatchRequests.receive()).isNotNull
+        assertThat(fakeBackfilaClientServiceClient.runBatchRequests.receive()).isNotNull
+        assertThat(fakeBackfilaClientServiceClient.runBatchRequests.receive()).isNotNull
+
+        waitForOtherCoroutines()
+        assertThat(fakeBackfilaClientServiceClient.runBatchRequests.poll()).isNull()
+      } finally {
+        runner.stop()
+      }
+    }
+  }
+
+  private suspend fun waitForOtherCoroutines() {
```

Comment:
I don’t love this. Good enough for now, but it feels bad

Reply:
Yeah agreed. It would have to be a separate refactor to get at the underlying coroutine scope.